### PR TITLE
Updating ProblemDetails docs

### PR DESCRIPTION
**Pull request by brunolins16**
Preview URL:

- [General](https://review.learn.microsoft.com/en-us/aspnet/core/fundamentals/error-handling?view=aspnetcore-7.0&branch=pr-en-us-27400#problem-details)
- [Minimal APIs](https://review.learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis/handle-errrors?view=aspnetcore-7.0&branch=pr-en-us-27400) ***new**
- [MVC](https://review.learn.microsoft.com/en-us/aspnet/core/web-api/handle-errors?view=aspnetcore-7.0&branch=pr-en-us-27400#default-problem-details-response)

### CODE REVIEW

**Comment by brunolins16** [comment]
@Rick-Anderson I just started working on it but since I need to add more Problem Details docs Minimal API and I don't want to redirect to MVC doc I am suggesting move part of the content to fundamentals. Open to different ideas.

File: aspnetcore/fundamentals/minimal-apis/handle-errrors.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+In non-development environments, use [Exception Handler Middleware](xref:fundamentals/error-handling#exception-handler-page) to produce an error payload. To configure the `Exception Handler Middleware` in the preceding Minimal API app, call <xref:Microsoft.AspNetCore.Builder.ExceptionHandlerExtensions.UseExceptionHandler%2A>. 
+
+For example, the following code change the app to respond an [RFC 7807](https://tools.ietf.org/html/rfc7807)-compliant payload to the client. For more information, see [Problem Details](#problem-details) section.
+
+
+``` csharp
+var builder = WebApplication.CreateBuilder(args);
+var app = builder.Build();
+
+app.UseExceptionHandler(exceptionHandlerApp
```
Comment (brunolins16) [inline aspnetcore/fundamentals/minimal-apis/handle-errrors.md:82]: @tdykstra @Rick-Anderson All sample codes here are very simple, so, I decided to add them directly. Do you know if there is any possibility to highlight lines in this case? If not, maybe I will move them to a small project to have it.

Comment (Rick-Anderson) [reply aspnetcore/fundamentals/minimal-apis/handle-errrors.md:82]: > Do you know if there is any possibility to highlight lines in this case? If not, maybe I will move them to a small project to have it.

Not at the moment. If you want put `<!-- comments in the MD file where it should be highlighted -->` we can create the sample and do the migration for you.

File: aspnetcore/fundamentals/error-handling/samples/7.x/ErrorHandlingSample/ErrorHandlingSample.csproj [hunk 2]
```diff
@@ -0,0 +1,13 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
```
[the remark this answers is not in the document]
Comment (Rick-Anderson) [reply aspnetcore/fundamentals/error-handling/samples/7.x/ErrorHandlingSample/ErrorHandlingSample.csproj:1]: > @Rick-Anderson I don't see an error handling project in the `Samples` repo. Do you want I clean it up and move this project to there?

That's the best but OK to leave it in this repo if we don't expect lots of customers to want to download the sample. Code clean up always appreciated.

File: aspnetcore/fundamentals/error-handling.md [hunk 3]
```diff
@@ -10,7 +10,310 @@ uid: fundamentals/error-handling
...
+
+:::code language="csharp" source="error-handling/samples/7.x/ErrorHandlingSample/Snippets/Program.cs" id="snippet_AddProblemDetails" highlight="3":::
+
+The next section shows how to customize the problem details response body.
+
+<a name="cpd7"></a>
+
+### Customizing problem details
+
+TODO
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/error-handling.md:236]: Per style guide, avoid ing -> Customize, not Customizing

```suggestion
<!--
### Customize problem details
See https://github.com/dotnet/AspNetCore.Docs/issues/27416
TODO
-->
```

File: aspnetcore/fundamentals/error-handling.md [hunk 4]
```diff
@@ -10,7 +10,310 @@ uid: fundamentals/error-handling
...
+
+#### `CustomizeProblemDetails` operation
+
+The generated problem details can be customized using <xref:Microsoft.AspNetCore.Http.ProblemDetailsOptions.CustomizeProblemDetails> and are applied to all auto-generated problem details.
+
+The following code uses <xref:Microsoft.AspNetCore.Http.ProblemDetailsOptions> to set <xref:Microsoft.AspNetCore.Http.ProblemDetailsOptions.CustomizeProblemDetails>:
+
+:::code language="csharp" source="error-handling/samples/7.x/ErrorHandlingSample/Snippets/Program.cs" id="snippet_CustomizeProblemDetails" highlight="3-5":::
+
+For example, a `HTTP Status 400 Bad Request` endpoint result produces the following problem details response body
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/error-handling.md:246]:
```suggestion
For example, an [`HTTP Status 400 Bad Request`](https://developer.mozilla.org/docs/Web/HTTP/Status/400) endpoint result produces the following problem details response body:
```

File: aspnetcore/fundamentals/error-handling.md [hunk 5]
```diff
@@ -10,7 +10,310 @@ uid: fundamentals/error-handling
...
+
+#### Custom `IProblemDetailsWriter`
+
+An <xref:Microsoft.AspNetCore.Http.IProblemDetailsWriter> implementation can be created for advanced customizations:
+
+:::code language="csharp" source="~/../AspNetCore.Docs.Samples/fundamentals/middleware/problem-details-service/SampleProblemDetailsWriter.cs" :::
+
+#### Problem details from Middleware
+
+An alternative approach to using <xref:Microsoft.AspNetCore.Http.ProblemDetailsOptions> to set <xref:Microsoft.AspNetCore.Http.ProblemDetailsOptions.CustomizeProblemDetails> is to set the <xref:Microsoft.AspNetCore.Http.ProblemDetailsContext.ProblemDetails> in middleware. A problem details response can be written by calling [`IProblemDetailsService.WriteAsync`](/dotnet/api/microsoft.aspnetcore.http.iproblemdetailsservice.writeasync?view=aspnetcore-7.0&preserve-view=true):
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/error-handling.md:265]:
```suggestion
An alternative approach to using <xref:Microsoft.AspNetCore.Http.ProblemDetailsOptions> with <xref:Microsoft.AspNetCore.Http.ProblemDetailsOptions.CustomizeProblemDetails> is to set the <xref:Microsoft.AspNetCore.Http.ProblemDetailsContext.ProblemDetails> in middleware. A problem details response can be written by calling [`IProblemDetailsService.WriteAsync`](/dotnet/api/microsoft.aspnetcore.http.iproblemdetailsservice.writeasync?view=aspnetcore-7.0&preserve-view=true):
```

File: aspnetcore/fundamentals/error-handling.md [hunk 6]
```diff
@@ -10,7 +10,310 @@ uid: fundamentals/error-handling
...
+
+:::code language="csharp" source="~/../AspNetCore.Docs.Samples/fundamentals/middleware/problem-details-service/Program.cs" id="snippet_middleware" highlight="5,19-40":::
+
+In the preceding code, the minimal API endpoints `/divide` and `/squareroot` return the expected custom problem response on error input.
+
+The API controller endpoints return the default problem response on error input, not the custom problem response. The default problem response is returned because the API controller has written to the response stream, [Problem details for error status codes](/aspnet/core/web-api/#problem-details-for-error-status-codes-1), before [`IProblemDetailsService.WriteAsync`](https://github.com/dotnet/aspnetcore/blob/ce2db7ea0b161fc5eb35710fca6feeafeeac37bc/src/Http/Http.Extensions/src/ProblemDetailsService.cs#L24) is called and the response will not be written again.
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/error-handling.md:271]:
```suggestion
The API controller endpoints return the default problem response on error input, not the custom problem response. The default problem response is returned because the API controller has written to the response stream, [Problem details for error status codes](/aspnet/core/web-api/#problem-details-for-error-status-codes-1), before [`IProblemDetailsService.WriteAsync`](https://github.com/dotnet/aspnetcore/blob/ce2db7ea0b161fc5eb35710fca6feeafeeac37bc/src/Http/Http.Extensions/src/ProblemDetailsService.cs#L24) is called and the response is **not** written again.
```

File: aspnetcore/fundamentals/error-handling.md [hunk 7]
```diff
@@ -10,7 +10,310 @@ uid: fundamentals/error-handling
...
+
+```json
+{
+"type":"https://tools.ietf.org/html/rfc7231#section-6.6.1",
+"title":"An error occurred while processing your request.",
+"status":500,"traceId":"00-b644<snip>-00"
+}
+```
+
+For most apps, the preceding code is all that for exceptions, however, the following section shows how get more detailed problem responses.
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/error-handling.md:297]:
```suggestion
For most apps, the preceding code is all that's needed for exceptions, however, the following section shows how get more detailed problem responses.
```

**Review by Rick-Anderson — APPROVED** [review]

File: aspnetcore/fundamentals/error-handling.md (hunk 6 above)
Comment (BrennanConroy) [reply aspnetcore/fundamentals/error-handling.md:271]: This section is hard to parse.

Rough suggestion:
"Controllers ignore customized problem details on error status codes due to (link). In order to customize problem details for Controllers (blah)."`

File: aspnetcore/fundamentals/error-handling.md [hunk 8]
```diff
@@ -10,7 +10,310 @@ uid: fundamentals/error-handling
...
+
+:::code language="csharp" source="~/../AspNetCore.Docs.Samples/fundamentals/middleware/problem-details-service/Controllers/ValuesController.cs" id="snippet3"  highlight="16-21":::
+
+## Produce a ProblemDetails payload for exceptions
+
+Consider the following app:
+
+:::code language="csharp" source="~/../AspNetCore.Docs.Samples/fundamentals/middleware/problem-details-service/Program.cs" id="snippet_apishort" highlight="4,8":::
+
+In non-development environments, when an exception occurs, the following a standardized [ProblemDetails response](https://datatracker.ietf.org/doc/html/rfc7807) is returned to the client:
```
Comment (BrennanConroy) [inline aspnetcore/fundamentals/error-handling.md:287]:
```suggestion
In non-development environments, when an exception occurs, the following is a standard [ProblemDetails response](https://datatracker.ietf.org/doc/html/rfc7807) that is returned to the client:
```

File: aspnetcore/fundamentals/error-handling.md (hunk 7 above)
Comment (BrennanConroy) [reply aspnetcore/fundamentals/error-handling.md:297]:
```suggestion
For most apps, the preceding code is all that's needed for exceptions, however, the following section shows how to get more detailed problem responses.
```

File: aspnetcore/fundamentals/minimal-apis/handle-errrors.md [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+uid: fundamentals/minimal-apis/handle-errors
+---
+
+# How to handle errors in Minimal API apps
+
+This article describes how to handle errors in Minimal API apps.
+
+## Exceptions
+
+In a Minimal API app, there are two different built-in centralized mechanism to handle unhandled exceptions:
```
Comment (BrennanConroy) [inline aspnetcore/fundamentals/minimal-apis/handle-errrors.md:17]:
```suggestion
In a Minimal API app, there are two different built-in centralized mechanisms to handle unhandled exceptions:
```

File: aspnetcore/fundamentals/minimal-apis/handle-errrors.md [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+```
+
+### Developer Exception Page
+
+The [Developer Exception Page](xref:fundamentals/error-handling#developer-exception-page) shows detailed stack traces for server errors. It uses <xref:Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware> to capture synchronous and asynchronous exceptions from the HTTP pipeline and to generate error responses. 
+
+ASP.NET Core apps enable the developer exception page by default when both:
+
+* Running in the [Development environment](xref:fundamentals/environments).
+* App created with the current templates, that is, using [WebApplication.CreateBuilder](/dotnet/api/microsoft.aspnetcore.builder.webapplication.createbuilder).
```
Comment (BrennanConroy) [inline aspnetcore/fundamentals/minimal-apis/handle-errrors.md:41]:
```suggestion
* App is using [WebApplication.CreateBuilder](/dotnet/api/microsoft.aspnetcore.builder.webapplication.createbuilder).
```

File: aspnetcore/fundamentals/minimal-apis/handle-errrors.md [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+Host: localhost:5239
+Accept-Encoding: gzip, deflate, br
+```
+
+> [!WARNING]
+> Don't enable the Developer Exception Page **unless the app is running in the Development environment**. Don't share detailed exception information publicly when the app runs in production. For more information on configuring environments, see <xref:fundamentals/environments>.
+
+### Exception handler
+
+In non-development environments, use [Exception Handler Middleware](xref:fundamentals/error-handling#exception-handler-page) to produce an error payload. To configure the `Exception Handler Middleware` in the preceding Minimal API app, call <xref:Microsoft.AspNetCore.Builder.ExceptionHandlerExtensions.UseExceptionHandler%2A>.
```
Comment (BrennanConroy) [inline aspnetcore/fundamentals/minimal-apis/handle-errrors.md:73]:
```suggestion
In non-development environments, use the [Exception Handler Middleware](xref:fundamentals/error-handling#exception-handler-page) to produce an error payload. To configure the `Exception Handler Middleware`, call <xref:Microsoft.AspNetCore.Builder.ExceptionHandlerExtensions.UseExceptionHandler%2A>. 
```

File: aspnetcore/fundamentals/minimal-apis/handle-errrors.md [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+```
+
+> [!WARNING]
+> Don't enable the Developer Exception Page **unless the app is running in the Development environment**. Don't share detailed exception information publicly when the app runs in production. For more information on configuring environments, see <xref:fundamentals/environments>.
+
+### Exception handler
+
+In non-development environments, use [Exception Handler Middleware](xref:fundamentals/error-handling#exception-handler-page) to produce an error payload. To configure the `Exception Handler Middleware` in the preceding Minimal API app, call <xref:Microsoft.AspNetCore.Builder.ExceptionHandlerExtensions.UseExceptionHandler%2A>. 
+
+For example, the following code change the app to respond an [RFC 7807](https://tools.ietf.org/html/rfc7807)-compliant payload to the client. For more information, see [Problem Details](#problem-details) section.
```
Comment (BrennanConroy) [inline aspnetcore/fundamentals/minimal-apis/handle-errrors.md:75]:
```suggestion
For example, the following code changes the app to respond with an [RFC 7807](https://tools.ietf.org/html/rfc7807)-compliant payload to the client. For more information, see [Problem Details](#problem-details) section.
```

File: aspnetcore/fundamentals/minimal-apis/handle-errrors.md [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+
+app.Map("/users/{id:int}", (int id) 
+    => id <= 0 ? Results.BadRequest() : Results.Ok(new User(id)) );
+
+app.Run();
+
+public record User(int Id);
+```
+
+The `/users` endpoint produces `200 OK` with a `json` representation of `User` when `id` greater than `0`, or, `400 BAD REQUEST` status code without a response body. For more information about creating response, see [Create responses in Minimal API apps](/aspnet/core/fundamentals/minimal-apis/responses).
```
Comment (BrennanConroy) [inline aspnetcore/fundamentals/minimal-apis/handle-errrors.md:109]:
```suggestion
The `/users` endpoint produces `200 OK` with a `json` representation of `User` when `id` is greater than `0`, otherwise a `400 BAD REQUEST` status code without a response body. For more information about creating response, see [Create responses in Minimal API apps](/aspnet/core/fundamentals/minimal-apis/responses).
```

File: aspnetcore/fundamentals/minimal-apis/handle-errrors.md [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+
+The [`Status Code Pages middleware`](xref:fundamentals/error-handling#sestatuscodepages) can be configure to produce a common body content, **when empty**, for all HTTP client (`400`-`499`) or server (`500` -`599`) responses. The middleware is configured by calling 
+[UseStatusCodePages](<xref:Microsoft.AspNetCore.Builder.StatusCodePagesExtensions.UseStatusCodePages%2A>) extension method.
+
+For example, the following example change the app to respond an [RFC 7807](https://tools.ietf.org/html/rfc7807)-compliant payload to the client for all client and server responses, including routing errors (eg. `404 NOT FOUND`). For more information, see [Problem Details](#problem-details) section.
```
Comment (BrennanConroy) [inline aspnetcore/fundamentals/minimal-apis/handle-errrors.md:114]:
```suggestion
For example, the following example changes the app to respond with an [RFC 7807](https://tools.ietf.org/html/rfc7807)-compliant payload to the client for all client and server responses, including routing errors (eg. `404 NOT FOUND`). For more information, see [Problem Details](#problem-details) section.
```

File: aspnetcore/fundamentals/minimal-apis/handle-errrors.md [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+    => id <= 0 ? Results.BadRequest() : Results.Ok(new User(id)) );
+
+app.Run();
+
+public record User(int Id);
+```
+
+## Problem details
+
+[Problem Details](https://www.rfc-editor.org/rfc/rfc7807.html) are not the only response format to describe a HTTP API error, however, they are commonly used to report errors for HTTP APIs.
```
Comment (BrennanConroy) [inline aspnetcore/fundamentals/minimal-apis/handle-errrors.md:134]: This is the same as the text on the other page, there is probably some way to share the content so it only needs to be written once.

File: aspnetcore/fundamentals/minimal-apis/handle-errrors.md [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+### Developer Exception Page
+
+The [Developer Exception Page](xref:fundamentals/error-handling#developer-exception-page) shows detailed stack traces for server errors. It uses <xref:Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware> to capture synchronous and asynchronous exceptions from the HTTP pipeline and to generate error responses. 
+
+ASP.NET Core apps enable the developer exception page by default when both:
+
+* Running in the [Development environment](xref:fundamentals/environments).
+* App created with the current templates, that is, using [WebApplication.CreateBuilder](/dotnet/api/microsoft.aspnetcore.builder.webapplication.createbuilder).
+
+For more information on configuring middleware see, [Middleware in Minimal API apps](/aspnet/core/fundamentals/minimal-apis/middleware).
```
Comment (tdykstra) [inline aspnetcore/fundamentals/minimal-apis/handle-errrors.md:43]:
```suggestion
For more information on configuring middleware, see [Middleware in Minimal API apps](/aspnet/core/fundamentals/minimal-apis/middleware).
```

File: aspnetcore/fundamentals/minimal-apis/handle-errrors.md [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+    => id <= 0 ? Results.BadRequest() : Results.Ok(new User(id)) );
+
+app.Run();
+
+public record User(int Id);
+```
+
+The `/users` endpoint produces `200 OK` with a `json` representation of `User` when `id` greater than `0`, or, `400 BAD REQUEST` status code without a response body. For more information about creating response, see [Create responses in Minimal API apps](/aspnet/core/fundamentals/minimal-apis/responses).
+
+The [`Status Code Pages middleware`](xref:fundamentals/error-handling#sestatuscodepages) can be configure to produce a common body content, **when empty**, for all HTTP client (`400`-`499`) or server (`500` -`599`) responses. The middleware is configured by calling
```
Comment (tdykstra) [inline aspnetcore/fundamentals/minimal-apis/handle-errrors.md:111]:
```suggestion
The [`Status Code Pages middleware`](xref:fundamentals/error-handling#sestatuscodepages) can be configured to produce a common body content, **when empty**, for all HTTP client (`400`-`499`) or server (`500` -`599`) responses. The middleware is configured by calling the 
```

File: aspnetcore/fundamentals/minimal-apis/handle-errrors.md [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+app.Run();
+
+public record User(int Id);
+```
+
+## Problem details
+
+[Problem Details](https://www.rfc-editor.org/rfc/rfc7807.html) are not the only response format to describe a HTTP API error, however, they are commonly used to report errors for HTTP APIs.
+
+Minimal API apps can be configured to generate problem details response for all HTTP client and server error responses that ***don't have a body content yet*** using the [`AddProblemDetails`](/dotnet/api/microsoft.extensions.dependencyinjection.problemdetailsservicecollectionextensions.addproblemdetails?view=aspnetcore-7.0&preserve-view=true) extension method on <xref:Microsoft.Extensions.DependencyInjection.IServiceCollection>.
```
Comment (tdykstra) [inline aspnetcore/fundamentals/minimal-apis/handle-errrors.md:136]:
```suggestion
Minimal API apps can be configured to generate problem details response for all HTTP client and server error responses that ***don't have a body content yet*** by using the [`AddProblemDetails`](/dotnet/api/microsoft.extensions.dependencyinjection.problemdetailsservicecollectionextensions.addproblemdetails?view=aspnetcore-7.0&preserve-view=true) extension method on <xref:Microsoft.Extensions.DependencyInjection.IServiceCollection>.
```

File: aspnetcore/fundamentals/minimal-apis/handle-errrors.md [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+
+The following middleware generates problem details HTTP responses when [`AddProblemDetails`](/dotnet/api/microsoft.extensions.dependencyinjection.problemdetailsservicecollectionextensions.addproblemdetails?view=aspnetcore-7.0&preserve-view=true) is called, except when not accepted by the client:
+
+* <xref:Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware>: Generates a problem details response when a custom handler is not defined.
+* <xref:Microsoft.AspNetCore.Diagnostics.StatusCodePagesMiddleware>: Generates a problem details response by default.
+* <xref:Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware>: Generates a problem details response in development when `text/html` is not accepted.
+
+
+The following code configures the app to generate a problem details:
```
Comment (tdykstra) [inline aspnetcore/fundamentals/minimal-apis/handle-errrors.md:145]:
```suggestion
The following code configures the app to generate problem details:
```

File: aspnetcore/web-api/handle-errors.md [hunk 20]
```diff
@@ -210,9 +202,7 @@ The automatic creation of a `ProblemDetails` for error status codes is disabled
 
 Using the preceding code, when an API controller returns `BadRequest`, an [HTTP 400](https://developer.mozilla.org/docs/Web/HTTP/Status/400) response status is returned with no response body. `SuppressMapClientErrors` prevents a `ProblemDetails` response from being created, even when calling `WriteAsync` for an API Controller endpoint. `WriteAsync` is explained later in this article.
 
-The next section shows how to customize the problem details response body to return a more helpful response.
-
-<a name="cpd7"></a>
+The next section shows how to customize the problem details response body, using <xref:Microsoft.AspNetCore.Http.ProblemDetailsOptions.CustomizeProblemDetails>, to return a more helpful response. For more customization options, see the [Customizing problem details](/aspnet/core/fundamentals/error-handling#cpd7).
```
Comment (tdykstra) [inline aspnetcore/web-api/handle-errors.md:205]:
```suggestion
The next section shows how to customize the problem details response body, using <xref:Microsoft.AspNetCore.Http.ProblemDetailsOptions.CustomizeProblemDetails>, to return a more helpful response. For more customization options, see [Customizing problem details](/aspnet/core/fundamentals/error-handling#cpd7).
```

File: aspnetcore/fundamentals/error-handling.md [hunk 21]
```diff
@@ -10,7 +10,310 @@ uid: fundamentals/error-handling
...
+
+<a name="pds7"></a>
+
+## Problem details
+
+[Problem Details](https://www.rfc-editor.org/rfc/rfc7807.html) are not the only response format to describe a HTTP API error, however, they are commonly used to report errors for HTTP APIs.
+
+The problem details service implements the <xref:Microsoft.AspNetCore.Http.IProblemDetailsService> interface, which supports creating problem details in ASP.NET Core. The [`AddProblemDetails`](/dotnet/api/microsoft.extensions.dependencyinjection.problemdetailsservicecollectionextensions.addproblemdetails?view=aspnetcore-7.0&preserve-view=true) extension method on <xref:Microsoft.Extensions.DependencyInjection.IServiceCollection> register the default `IProblemDetailsService` implementation.
+
+In ASP.NET Core apps, the following middleware generates problem details HTTP responses when [`AddProblemDetails`](/dotnet/api/microsoft.extensions.dependencyinjection.problemdetailsservicecollectionextensions.addproblemdetails?view=aspnetcore-7.0&preserve-view=true) is called, except when not accepted by the client:
```
[the remark this answers is not in the document]
Comment (brunolins16) [reply aspnetcore/fundamentals/error-handling.md:220]: Basically, when the `Accepts` does not include one of the expected values. Eg.:https://source.dot.net/#Microsoft.AspNetCore.Http.Extensions/DefaultProblemDetailsWriter.cs,28

File: aspnetcore/fundamentals/minimal-apis/handle-errrors.md [hunk 22]
```diff
@@ -0,0 +1,165 @@
...
+This article describes how to handle errors in Minimal API apps.
+
+## Exceptions
+
+In a Minimal API app, there are two different built-in centralized mechanism to handle unhandled exceptions:
+
+* [Developer Exception Page middleware](#developer-exception-page)  **Development environment only**
+* [Exception handler middleware](#exception-handler)
+
+Consider the following Minimal API app, which throws an exception when the endpoint `/exception` is requested:
```
Comment (brunolins16) [inline aspnetcore/fundamentals/minimal-apis/handle-errrors.md:22]:
```suggestion
This section refers to the following Minimal API app to demonstrate ways to handle exceptions. It throws an exception when the endpoint `/exception` is requested:
```

**Review by tdykstra — COMMENTED** [review]
We can link to the new [error-handling in minimal APIs doc](https://review.learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis/handle-errrors?view=aspnetcore-7.0&branch=pr-en-us-27400) from the [Developer exception page section of the Minimal API overview doc](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis?view=aspnetcore-7.0#developer-exception-page-1).

File: aspnetcore/fundamentals/error-handling.md [hunk 23]
```diff
@@ -10,7 +10,310 @@ uid: fundamentals/error-handling
...
+For most apps, the preceding code is all that for exceptions, however, the following section shows how get more detailed problem responses.
+
+An alternative to a [custom exception handler page](xref:fundamentals/error-handling#exception-handler-page) is to provide a lambda to <xref:Microsoft.AspNetCore.Builder.ExceptionHandlerExtensions.UseExceptionHandler%2A>. Using a lambda allows access to the error and writing a problem details response with [`IProblemDetailsService.WriteAsync`](/dotnet/api/microsoft.aspnetcore.http.iproblemdetailsservice.writeasync?view=aspnetcore-7.0&preserve-view=true):
+
+:::code language="csharp" source="~/../AspNetCore.Docs.Samples/fundamentals/middleware/problem-details-service/Program.cs" id="snippet_lambda" :::
+
+> [!WARNING]
+> Do **not** serve sensitive error information to clients. Serving errors is a security risk.
+
+[Hellang.Middleware.ProblemDetails](https://www.nuget.org/packages/Hellang.Middleware.ProblemDetails/) is a 3rd party problem details middleware Nuget package.
```
[the remark this answers is not in the document]
Comment (brunolins16) [reply aspnetcore/fundamentals/error-handling.md]: I add additional information. Let me know if good enough.

File: aspnetcore/fundamentals/error-handling.md [hunk 24]
```diff
@@ -10,7 +10,314 @@ uid: fundamentals/error-handling
...
+For most apps, the preceding code is all that's needed for exceptions, however, the following section shows how to get more detailed problem responses.
+
+An alternative to a [custom exception handler page](xref:fundamentals/error-handling#exception-handler-page) is to provide a lambda to <xref:Microsoft.AspNetCore.Builder.ExceptionHandlerExtensions.UseExceptionHandler%2A>. Using a lambda allows access to the error and writing a problem details response with [`IProblemDetailsService.WriteAsync`](/dotnet/api/microsoft.aspnetcore.http.iproblemdetailsservice.writeasync?view=aspnetcore-7.0&preserve-view=true):
+
+:::code language="csharp" source="~/../AspNetCore.Docs.Samples/fundamentals/middleware/problem-details-service/Program.cs" id="snippet_lambda" :::
+
+> [!WARNING]
+> Do **not** serve sensitive error information to clients. Serving errors is a security risk.
+
+An alternative approach to generate problem details is to use the 3rd party Nuget package [Hellang.Middleware.ProblemDetails](https://www.nuget.org/packages/Hellang.Middleware.ProblemDetails/) that can be use to map exceptions and client errors to a problem details.
```
Comment (tdykstra) [inline aspnetcore/fundamentals/error-handling.md:310]:
```suggestion
An alternative approach to generate problem details is to use the 3rd party Nuget package [Hellang.Middleware.ProblemDetails](https://www.nuget.org/packages/Hellang.Middleware.ProblemDetails/) that can be used to map exceptions and client errors to problem details.
```

File: aspnetcore/fundamentals/error-handling.md (hunk 23 above)
[the remark this answers is not in the document]
Comment (tdykstra) [reply aspnetcore/fundamentals/error-handling.md]: Looks good, thanks.

File: aspnetcore/fundamentals/error-handling.md (hunk 21 above)
[the remark this answers is not in the document]
Comment (brunolins16) [reply aspnetcore/fundamentals/error-handling.md]: Is that better now?

File: aspnetcore/fundamentals/error-handling.md [hunk 25]
```diff
@@ -10,7 +10,310 @@ uid: fundamentals/error-handling
...
+
+In ASP.NET Core apps, the following middleware generates problem details HTTP responses when [`AddProblemDetails`](/dotnet/api/microsoft.extensions.dependencyinjection.problemdetailsservicecollectionextensions.addproblemdetails?view=aspnetcore-7.0&preserve-view=true) is called, except when not accepted by the client:
+
+* <xref:Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware>: Generates a problem details response when a custom handler is not defined.
+* <xref:Microsoft.AspNetCore.Diagnostics.StatusCodePagesMiddleware>: Generates a problem details response by default.
+* <xref:Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware>: Generates a problem details response in development when `text/html` is not accepted.
```
[the remark this answers is not in the document]
Comment (brunolins16) [reply aspnetcore/fundamentals/error-handling.md]: Done

File: aspnetcore/fundamentals/error-handling.md (hunk 21 above)
[the remark this answers is not in the document]
Comment (BrennanConroy) [reply aspnetcore/fundamentals/error-handling.md]: > except when the [`Accept` request HTTP header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept) doesn't include one of the content types supported by the registered <xref:Microsoft.AspNetCore.Http.IProblemDetailsWriter> (default: `application/json`):

Yes 😃

**Review by BrennanConroy — APPROVED** [review]

**Comment by brunolins16** [comment]
> We can link to the new [error-handling in minimal APIs doc](https://review.learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis/handle-errrors?view=aspnetcore-7.0&branch=pr-en-us-27400) from the [Developer exception page section of the Minimal API overview doc](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis?view=aspnetcore-7.0#developer-exception-page-1).

@tdykstra I will skip this until we decide how the overview page will be, is that ok?

**Comment by Rick-Anderson** [comment]
@brunolins16  yes, OK if I S&M and make live?

**Comment by tdykstra** [comment]
>I will skip this until we decide how the overview page will be, is that ok?

Yes, definitely.
